tests/cecil-tests/Helper.cs

```diff
@@ -219,6 +224,139 @@ public static IEnumerable<ICustomAttributeProvider> EnumerateAttributeProviders
+				yield return item;
+
+			foreach (var item in assembly.EnumerateMethods (filter))
+				yield return item;
+
+			foreach (var item in assembly.EnumerateProperties (filter))
+				yield return item;
+
+			foreach (var item in assembly.EnumerateEvents (filter))
+				yield return item;
+		}
+
+		public static bool IsPubliclyVisible (this TypeDefinition type)
+		{
+			if (type.IsNested) {
+				if (type.IsNestedAssembly || type.IsNestedFamilyAndAssembly || type.IsNestedPrivate)
+					return false;
+				return IsPubliclyVisible (type.DeclaringType);
+			}
+
+			return type.IsPublic;
+		}
+
+		public static bool IsPubliclyVisible (this FieldDefinition field)
+		{
+			if (!IsPubliclyVisible (field.DeclaringType))
+				return false;
+
+			var visibility = field.Attributes & FieldAttributes.FieldAccessMask;
+			switch (visibility) {
+			case FieldAttributes.Private:
+			case FieldAttributes.FamANDAssem:
+			case FieldAttributes.Assembly:
+				return false;
+			case FieldAttributes.Family:
+			case FieldAttributes.FamORAssem:
+			case FieldAttributes.Public:
+				return true;
+			default:
+				throw new NotImplementedException ($"Unknown visibility: {visibility}");
+			}
+		}
+
+		public static bool IsPubliclyVisible (this MethodDefinition method)
+		{
+			if (!IsPubliclyVisible (method.DeclaringType))
+				return false;
+
+			var visibility = method.Attributes & MethodAttributes.MemberAccessMask;
+			switch (visibility) {
+			case MethodAttributes.Private:
+			case MethodAttributes.FamANDAssem:
+			case MethodAttributes.Assembly:
+				return false;
+			case MethodAttributes.Family:
+			case MethodAttributes.FamORAssem:
+			case MethodAttributes.Public:
+				return true;
+			default:
+				throw new NotImplementedException ($"Unknown visibility: {visibility}");
+			}
+		}
+
+		public static bool IsPubliclyVisible (this EventDefinition evt)
+		{
+			if (!IsPubliclyVisible (evt.DeclaringType))
+				return false;
+
+			var invokeMethod = evt.InvokeMethod;
+			if (invokeMethod is not null && IsPubliclyVisible (invokeMethod))
+				return true;
+			var addMethod = evt.AddMethod;
+			if (addMethod is not null && IsPubliclyVisible (addMethod))
+				return true;
+			var removeMethod = evt.RemoveMethod;
+			if (removeMethod is not null && IsPubliclyVisible (removeMethod))
+				return true;
+			return false;
+		}
+
+		public static bool IsPubliclyVisible (this PropertyDefinition property)
+		{
+			if (!IsPubliclyVisible (property.DeclaringType))
+				return false;
+
+			var getter = property.GetMethod;
+			if (getter is not null && IsPubliclyVisible (getter))
+				return true;
+			var setter = property.SetMethod;
+			if (setter is not null && IsPubliclyVisible (setter))
+				return true;
+			return false;
+		}
+
+		public static IEnumerable<MemberReference> EnumeratePublicMembers (this AssemblyDefinition assembly, Func<MemberReference, bool>? filter = null)
+		{
+			foreach (var item in assembly.EnumerateTypes (filter)) {
```

Comment:
I think we could make things cleaner if we did a function composition with the filter and the IsPubliclyVisible functions to reduce the need to iterate more than needed. Both are Func<MemeberReference, bool> and we could combine them like this:

```csharp
public static Func<T, bool> Compose<T>(this Func<T, bool>? func1, Func<T, bool> func2)
{
	return func1 is null ? func2 : mr => func1 (mr) && func2 (mr);
}
```

Once we have the ability to compose the filter function with a similar one, we can create a IsVisible function like the following one:

```csharp
public static bool IsVisible (MemberReference memberReference) => memberReference switch {
		PropertyDefinition p => p.IsVisible (),
		MethodDefinition m => m.IsVisible (),
		_ => throw new ArgumentException ()
};
```
Taking those two functions + the Concate from Linq, this method could be:
```
public static IEnumerable<MemberReference> EnumeratePublicMembers (this AssemblyDefinition assembly, Func<MemberReference, bool>? filter = null)
{
	var compose = filter.Compose (IsVisible);
	return assembly.EnumerateProperties (compose)
		.Concat<MemberReference> (assembly.EnumerateMethods (compose));
}
```

Reply:
OK, simplified it somewhat.